File: sdk/go/spl/crypto_test.go
package spl

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

func loadVectors(t *testing.T, name string) map[string]any {
	t.Helper()
	data, err := os.ReadFile("../../../examples/crypto/" + name)
	if err != nil {
		t.Skipf("skipping crypto test: %v", err)
	}
	var v map[string]any
	if err := json.Unmarshal(data, &v); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	return v
}

func TestEd25519ValidSignature(t *testing.T) {
	v := loadVectors(t, "ed25519_vectors.json")
	msg := v["message"].(string)
	sig := v["signature_hex"].(string)
	pub := v["public_key_hex"].(string)

	if !VerifyEd25519([]byte(msg), sig, pub) {
		t.Fatal("expected valid signature to verify")
	}
}

func TestEd25519TamperedMessage(t *testing.T) {
	v := loadVectors(t, "ed25519_vectors.json")
	tampered := v["tampered_message"].(string)
	sig := v["signature_hex"].(string)
	pub := v["public_key_hex"].(string)

	if VerifyEd25519([]byte(tampered), sig, pub) {
		t.Fatal("expected tampered message to fail verification")
	}
}

func TestMerkleProofValid(t *testing.T) {
	v := loadVectors(t, "merkle_vectors.json")
	root := v["root"].(string)
	cases := v["cases"].([]any)

	for _, c := range cases {
		tc := c.(map[string]any)
		name := tc["name"].(string)
		leaf := tc["leaf"].(string)
		leafHashExpected := tc["leaf_hash"].(string)
		expected := tc["expected"].(bool)

		// Verify leaf hash
		h := sha256.Sum256([]byte(leaf))
		if hex.EncodeToString(h[:]) != leafHashExpected {
			t.Fatalf("%s: leaf hash mismatch", name)
		}

		proofRaw := tc["proof"].([]any)
		var proof []MerkleProofStep
		for _, p := range proofRaw {
			pm := p.(map[string]any)
			proof = append(proof, MerkleProofStep{
				Hash:     pm["hash"].(string),
				Position: pm["position"].(string),
			})
		}

		result := VerifyMerkleProof(leaf, proof, root)
		if result != expected {
			t.Fatalf("%s: expected %v, got %v", name, expected, result)
		}
	}
}

func TestHashChainValid(t *testing.T) {
	v := loadVectors(t, "hashchain_vectors.json")
	commitment := v["commitment"].(string)
	chainLength := int(v["chain_length"].(float64))
	cases := v["cases"].([]any)

	for _, c := range cases {
		tc := c.(map[string]any)
		name := tc["name"].(string)
		preimage := tc["preimage"].(string)
		index := int(tc["index"].(float64))
		expected := tc["expected"].(bool)

		result := VerifyHashChain(commitment, preimage, index, chainLength)
		if result != expected {
			t.Fatalf("%s: expected %v, got %v", name, expected, result)
		}
	}
}

File: sdk/js/src/crypto.test.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { verifyEd25519, verifyMerkleProof, verifyHashChain, sha256 } from './crypto.js';

function toHex(buf: Uint8Array): string {
  return Buffer.from(buf).toString('hex');
}

const vectorsDir = resolve(import.meta.dirname!, '../../../examples/crypto');

function loadVectors(name: string) {
  return JSON.parse(readFileSync(resolve(vectorsDir, name), 'utf8'));
}

describe('Ed25519 crypto', () => {
  const v = loadVectors('ed25519_vectors.json');

  it('verifies valid signature', () => {
    assert.equal(
      verifyEd25519(v.message, v.signature_hex, v.public_key_hex),
      true,
    );
  });

  it('rejects tampered message', () => {
    assert.equal(
      verifyEd25519(v.tampered_message, v.signature_hex, v.public_key_hex),
      false,
    );
  });
});

describe('Merkle proof', () => {
  const v = loadVectors('merkle_vectors.json');

  for (const tc of v.cases) {
    it(tc.name, () => {
      const leafHash = toHex(sha256(tc.leaf));
      assert.equal(leafHash, tc.leaf_hash);
      assert.equal(
        verifyMerkleProof(tc.leaf, tc.proof, v.root),
        tc.expected,
      );
    });
  }
});

describe('Hash chain', () => {
  const v = loadVectors('hashchain_vectors.json');

  for (const tc of v.cases) {
    it(tc.name, () => {
      assert.equal(
        verifyHashChain(v.commitment, tc.preimage, tc.index, v.chain_length),
        tc.expected,
      );
    });
  }
});
